File: solutions/Golang/average_of_levels_in_binary_tree_dfs.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func AverageOfLevels(root *TreeNode) []float64 {
	sums := []int64{}
	counts := []int{}

	var dfs func(*TreeNode, int)
	dfs = func(node *TreeNode, level int) {
		if node == nil {
			return
		}

		if level == len(sums) {
			sums = append(sums, 0)
			counts = append(counts, 0)
		}

		sums[level] += int64(node.Val)
		counts[level]++

		dfs(node.Left, level+1)
		dfs(node.Right, level+1)
	}

	dfs(root, 0)

	result := make([]float64, len(sums))
	for i := 0; i < len(sums); i++ {
		result[i] = float64(sums[i]) / float64(counts[i])
	}

	return result
}

func main() {
	root := &TreeNode{Val: 3}
	root.Left = &TreeNode{Val: 9}
	root.Right = &TreeNode{Val: 20}
	root.Right.Left = &TreeNode{Val: 15}
	root.Right.Right = &TreeNode{Val: 7}

	result := AverageOfLevels(root)
	fmt.Println("Averages:", result)
}

File: solutions/C++/average_of_levels_in_binary_tree_dfs.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
private:
    void dfs(TreeNode* node, int level, vector<long long>& sums, vector<int>& counts) {
        if (!node) return;

        if (level == sums.size()) {
            sums.push_back(0);
            counts.push_back(0);
        }

        sums[level] += node->val;
        counts[level]++;

        dfs(node->left, level + 1, sums, counts);
        dfs(node->right, level + 1, sums, counts);
    }

public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<long long> sums;
        vector<int> counts;
        dfs(root, 0, sums, counts);

        vector<double> result;
        for (int i = 0; i < sums.size(); i++) {
            result.push_back((double)sums[i] / counts[i]);
        }
        return result;
    }
};

int main() {
    Solution sol;
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);

    vector<double> result = sol.averageOfLevels(root);
    cout << "Averages: ";
    for (double val : result) cout << val << " ";
    cout << endl;

    return 0;
}
